Add SPI to report page load time

rniwa opened the pull request:
#### b6acbfad5859fb4175817208e3bcb6728d85ac32
<pre>
Add SPI to report page load time
<a href="https://bugs.webkit.org/show_bug.cgi?id=278554">https://bugs.webkit.org/show_bug.cgi?id=278554</a>

Reviewed by Chris Dumez.

Add new selector to WKNavigationDelegate which reports the key page load time metrics
such as when the first meaningful paint has happened, main frame&apos;s document had
finished loading (i.e. parsing), and when all subresources had finished loading using
_WKPageLoadTiming.

Subresources are assumed to be finished loading when all existing subresources had
finished loading and no new load was initiated within 100ms.

We use WallTime throughout instead of MonotonicTime since MonotonicTime doesn&apos;t
convert well to NSDate.

* Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKNavigationDelegatePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
* Source/WebKit/UIProcess/API/Cocoa/_WKPageLoadTiming.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKPageLoadTiming.mm: Added.
(nsDateFromMonotonicTime):
(-[_WKPageLoadTiming _initWithTiming:]):
(-[_WKPageLoadTiming navigationStart]):
(-[_WKPageLoadTiming firstMeaningfulPaint]):
(-[_WKPageLoadTiming documentFinishedLoading]):
(-[_WKPageLoadTiming allSubresourcesFinishedLoading]):
* Source/WebKit/UIProcess/API/Cocoa/_WKPageLoadTimingInternal.h: Added.
* Source/WebKit/UIProcess/Cocoa/CoreTelephonyUtilities.h:
* Source/WebKit/UIProcess/Cocoa/NavigationState.h:
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::NavigationState::didGeneratePageLoadTiming):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::didGeneratePageLoadTiming):
(WebKit::WebPageProxy::didCommitLayerTree):
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::didStartProvisionalLoadForFrame):
(WebKit::ProvisionalPageProxy::startNetworkRequestsForPageLoadTiming):
(WebKit::ProvisionalPageProxy::didReceiveMessage):
* Source/WebKit/UIProcess/ProvisionalPageProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction):
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::didStartProvisionalLoadForFrame):
* Source/WebKit/UIProcess/RemotePageProxy.h:
* Source/WebKit/UIProcess/WebPageLoadTiming.h: Added,
(WebKit::WebPageLoadTiming::WebPageLoadTiming):
(WebKit::WebPageLoadTiming::navigationStart const):
(WebKit::WebPageLoadTiming::firstMeaningfulPaint const):
(WebKit::WebPageLoadTiming::setFirstMeaningfulPaint):
(WebKit::WebPageLoadTiming::documentFinishedLoading const):
(WebKit::WebPageLoadTiming::setDocumentFinishedLoading):
(WebKit::WebPageLoadTiming::allSubresourcesFinishedLoading const):
(WebKit::WebPageLoadTiming::updateEndOfNetworkRequests):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::WebPageProxy):
(WebKit::WebPageProxy::startNetworkRequestsForPageLoadTiming):
(WebKit::WebPageProxy::endNetworkRequestsForPageLoadTiming):
(WebKit::WebPageProxy::generatePageLoadingTimingSoon):
(WebKit::WebPageProxy::didEndNetworkRequestsForPageLoadTimingTimerFired):
(WebKit::WebPageProxy::didStartProvisionalLoadForFrame):
(WebKit::WebPageProxy::didStartProvisionalLoadForFrameShared):
(WebKit::WebPageProxy::didFinishDocumentLoadForFrame):
(WebKit::WebPageProxy::didLayoutForCustomContentProvider):
(WebKit::WebPageProxy::didReachLayoutMilestone):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::useNetworkLoader):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::dispatchDidStartProvisionalLoad):
(WebKit::WebLocalFrameLoaderClient::dispatchDidFinishDocumentLoad):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):
(WebKit::WebPage::addResourceRequest):
(WebKit::WebPage::removeResourceRequest):
(WebKit::WebPage::dispatchDidReachLayoutMilestone):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/mac/TiledCoreAnimationDrawingArea.mm:
(WebKit::TiledCoreAnimationDrawingArea::sendPendingNewlyReachedPaintingMilestones):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(TEST(WKNavigation, GeneratePageLoadTiming)):
* Tools/TestWebKitAPI/cocoa/TestNavigationDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestNavigationDelegate.mm:
(-[TestNavigationDelegate _webView:didGeneratePageLoadTiming:]):

Canonical link: <a href="https://commits.webkit.org/282826@main">https://commits.webkit.org/282826@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a3afc0953ce8a8c9120cd3af522323ef3fc5f837

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64269 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43626 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16866 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68291 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14877 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/66389 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51324 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15157 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/51722 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10256 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67338 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40334 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55605 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32341 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37004 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12986 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13751 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/58967 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13316 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69990 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8216 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12834 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59041 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8249 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55699 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/59203 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6789 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/474 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9759 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39446 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/40525 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/41708 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40268 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKNavigationDelegatePrivate.h:
```diff
@@ -124,4 +125,6 @@ static const WKNavigationResponsePolicy _WKNavigationResponsePolicyBecomeDownloa
 - (void)_webView:(WKWebView *)webView willGoToBackForwardListItem:(WKBackForwardListItem *)item inPageCache:(BOOL)inPageCache WK_API_AVAILABLE(macos(10.13.4), ios(14.0));
 - (void)_webView:(WKWebView *)webView decidePolicyForSOAuthorizationLoadWithCurrentPolicy:(_WKSOAuthorizationLoadPolicy)policy forExtension:(NSString *)extension completionHandler:(void (^)(_WKSOAuthorizationLoadPolicy policy))completionHandler WK_API_AVAILABLE(macos(10.15), ios(13.0));
 
+- (void)_webView:(WKWebView *)webView didGeneratePageLoadTiming:(_WKPageLoadTiming*)timing WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
```
space before *

rniwa replied:
Fixed.

achristensen07 commented on Source/WebKit/UIProcess/Cocoa/CoreTelephonyUtilities.h:
```diff
@@ -25,6 +25,8 @@
 
 #pragma once
 
+#include <wtf/URL.h>
```
wtf/Forward.h should suffice.

rniwa replied:
Fixed.

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/_WKPageLoadTiming.mm:
```diff
@@ -0,0 +1,79 @@
+    if (!(self = [super init]))
+        return nil;
+
+    _navigationStart = timing.navigationStart();
```
Can we make this a new API::Object like we do with most of the rest of the API?  _WKResourceLoadInfo is a similar example.

rniwa replied:
Huh, funny you say that because I had that design initially and went back to this model for simplicity. Adding new API object seems like an unnecessary overhead there. Any reason you'd prefer that way?

achristensen07 commented on Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
```diff
@@ -1676,4 +1677,13 @@ static _WKSOAuthorizationLoadPolicy wkSOAuthorizationLoadPolicy(SOAuthorizationL
+    if (!delegate)
+        return;
+    RetainPtr wkPageLoadTiming = adoptNS([[_WKPageLoadTiming alloc] _initWithTiming:timing]);
+    [static_cast<id<WKNavigationDelegatePrivate>>(delegate) _webView:webView().get() didGeneratePageLoadTiming:wkPageLoadTiming.get()];
```
Then this would just call wrapper(timing)

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -6091,6 +6093,36 @@ void WebPageProxy::setNetworkRequestsInProgress(bool networkRequestsInProgress)
+void WebPageProxy::generatePageLoadingTimingSoon()
+{
+    m_generatePageLoadTimingTimer.stop();
+    if (!m_pageLoadTiming || m_subresourceLoadingCountForPageLoadTiming || !m_pageLoadTiming->firstMeaningfulPaint() || !m_pageLoadTiming->documentFinishedLoading() || !m_pageLoadTiming->allSubresourcesFinishedLoading())
```
Could each of these be on its own line?

rniwa replied:
Sure, we can do that.

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6612,6 +6612,14 @@ void WebPage::drawPagesForPrinting(FrameIdentifier frameID, const PrintInfo& pri
+        bool wasEmpty = m_networkResourceRequestIdentifiersForPageLoadTiming.isEmpty();
+        m_networkResourceRequestIdentifiersForPageLoadTiming.add(identifier);
+        if (wasEmpty)
+            send(Messages::WebPageProxy::StartNetworkRequestsForPageLoadTiming());
```
This function already sends Messages::WebPageProxy::SetNetworkRequestsInProgress below.  Could they be combined into one?

rniwa replied:
No, the exact condition under which these messages are sent are different.

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6627,6 +6635,11 @@ void WebPage::addResourceRequest(WebCore::ResourceLoaderIdentifier identifier, c
+    auto didRemove = m_networkResourceRequestIdentifiersForPageLoadTiming.remove(identifier);
+    ASSERT_UNUSED(didRemove, didRemove);
+    if (m_networkResourceRequestIdentifiersForPageLoadTiming.isEmpty())
+        send(Messages::WebPageProxy::EndNetworkRequestsForPageLoadTiming(WallTime::now()));
```
Ditto.

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -2600,6 +2600,7 @@ class WebPage : public API::ObjectImpl<API::Object::Type::BundlePage>, public IP
     };
     std::optional<DeferredDidReceiveMouseEvent> m_deferredDidReceiveMouseEvent;
 
+    HashSet<WebCore::ResourceLoaderIdentifier> m_networkResourceRequestIdentifiersForPageLoadTiming;
```
Maybe a separate PR, but this and m_trackedNetworkResourceRequestIdentifiers could be abstracted into an object that in release builds just increment and decrement an integer and in debug builds maintains a set to assert that we aren't re-adding an existing one.

achristensen07 reviewed (changes requested):
This should have a test to verify the delegate is called and the times are reasonable (sequential, after the time a WKWebView was allocated, etc.)

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKPageLoadTiming.mm:
```diff
@@ -0,0 +1,79 @@
+    WallTime _allSubresourcesFinishedLoading;
+}
+
+- (instancetype)_initWithTiming:(WebKit::WebPageLoadTiming &)timing
```
extra space before `&`? also, maybe the parameter can be `const` too?

rniwa replied:
Oh, I believe this is the convention for Objective-C code but will make it const.

cdumez replied:
I think this convention applies to Objective C types, not C++ types in a mm file.

cdumez replied:
This convention only applies to `*` as far as I know, not `&`.

rniwa replied:
Hm... ok.

cdumez commented on Source/WebKit/UIProcess/Cocoa/NavigationState.h:
```diff
@@ -283,6 +286,7 @@ class NavigationState final : public PageLoadState::Observer {
 #if HAVE(APP_SSO)
         bool webViewDecidePolicyForSOAuthorizationLoadWithCurrentPolicyForExtensionCompletionHandler : 1;
 #endif
+        bool didGeneratePageLoadTiming : 1;
```
I think this should start with `webView`, like the other ones, since the new method is `webView: something..`

rniwa replied:
Sure, will fix.

cdumez commented on Source/WebKit/UIProcess/WebPageLoadTiming.h:
```diff
@@ -0,0 +1,64 @@
+    WTF_MAKE_NONCOPYABLE(WebPageLoadTiming);
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    WebPageLoadTiming(WallTime navigationStart)
```
`explicit`

rniwa replied:
Sure.

cdumez commented on Source/WebKit/UIProcess/WebPageLoadTiming.h:
```diff
@@ -0,0 +1,64 @@
+    }
+
+private:
+    WallTime m_navigationStart;
```
Curious why are we using WallTime and not MonotonicTime? I'm not saying it's wrong but I'm curious of the reasoning. For example, I don't think we ever want those to move backwards if daylight savings hits in the middle of a page load?

rniwa replied:
I mentioned this in the commit message but basically MonotonicTime doesn't convert well to NSDate. `secondsSinceEpoch()` isn't really seconds since Jan 1st 1970.

cdumez replied:
I admit I failed to read the commit log :D

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -7009,16 +7059,24 @@ void WebPageProxy::didFirstVisuallyNonEmptyLayoutForFrame(IPC::Connection& conne
-void WebPageProxy::didReachLayoutMilestone(OptionSet<WebCore::LayoutMilestone> layoutMilestones)
+void WebPageProxy::didReachLayoutMilestone(OptionSet<WebCore::LayoutMilestone> layoutMilestones, WallTime timestamp)
 {
+    UNUSED_PARAM(timestamp);
```
Doesn't seem to be needed?

rniwa replied:
Oh, sure, will remove.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2572,6 +2577,14 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
+    void didEndNetworkRequestsForPageLoadTimingTimerFired();
+    void generatePageLoadingTimingSoon();
+#if PLATFORM(COCOA)
+    void didGeneratePageLoadTiming(WebPageLoadTiming&);
```
Can the parameter be const?

rniwa replied:
Sure, will do.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2572,6 +2577,14 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
+#if PLATFORM(COCOA)
+    void didGeneratePageLoadTiming(WebPageLoadTiming&);
+#else
+    void didGeneratePageLoadTiming(WebPageLoadTiming&) { }
```
ditto.

rniwa replied:
Ditto.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6620,6 +6621,14 @@ void WebPage::drawPagesForPrinting(FrameIdentifier frameID, const PrintInfo& pri
+        bool wasEmpty = m_networkResourceRequestIdentifiersForPageLoadTiming.isEmpty();
+        m_networkResourceRequestIdentifiersForPageLoadTiming.add(identifier);
+        if (wasEmpty)
+            send(Messages::WebPageProxy::StartNetworkRequestsForPageLoadTiming());
```
Theoretically, we can send this IPC before adding the identifier to the map and avoid the `wasEmpty` local variable.

I'm also unclear why you wrapped all this new code in a new scope using curly brackets.

rniwa replied:
Sure, will change the order.

cdumez reviewed (approved):
LGTM with nits.

rniwa committed:
Add SPI to report page load time
https://bugs.webkit.org/show_bug.cgi?id=278554

Reviewed by Chris Dumez.

Add new selector to WKNavigationDelegate which reports the key page load time metrics
such as when the first meaningful paint has happened, main frame's document had
finished loading (i.e. parsing), and when all subresources had finished loading using
_WKPageLoadTiming.

Subresources are assumed to be finished loading when all existing subresources had
finished loading and no new load was initiated within 100ms.

We use WallTime throughout instead of MonotonicTime since MonotonicTime doesn't
convert well to NSDate.

* Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKNavigationDelegatePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
* Source/WebKit/UIProcess/API/Cocoa/_WKPageLoadTiming.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKPageLoadTiming.mm: Added.
(nsDateFromMonotonicTime):
(-[_WKPageLoadTiming _initWithTiming:]):
(-[_WKPageLoadTiming navigationStart]):
(-[_WKPageLoadTiming firstMeaningfulPaint]):
(-[_WKPageLoadTiming documentFinishedLoading]):
(-[_WKPageLoadTiming allSubresourcesFinishedLoading]):
* Source/WebKit/UIProcess/API/Cocoa/_WKPageLoadTimingInternal.h: Added.
* Source/WebKit/UIProcess/Cocoa/CoreTelephonyUtilities.h:
* Source/WebKit/UIProcess/Cocoa/NavigationState.h:
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::NavigationState::didGeneratePageLoadTiming):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::didGeneratePageLoadTiming):
(WebKit::WebPageProxy::didCommitLayerTree):
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::didStartProvisionalLoadForFrame):
(WebKit::ProvisionalPageProxy::startNetworkRequestsForPageLoadTiming):
(WebKit::ProvisionalPageProxy::didReceiveMessage):
* Source/WebKit/UIProcess/ProvisionalPageProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction):
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::didStartProvisionalLoadForFrame):
* Source/WebKit/UIProcess/RemotePageProxy.h:
* Source/WebKit/UIProcess/WebPageLoadTiming.h: Added,
(WebKit::WebPageLoadTiming::WebPageLoadTiming):
(WebKit::WebPageLoadTiming::navigationStart const):
(WebKit::WebPageLoadTiming::firstMeaningfulPaint const):
(WebKit::WebPageLoadTiming::setFirstMeaningfulPaint):
(WebKit::WebPageLoadTiming::documentFinishedLoading const):
(WebKit::WebPageLoadTiming::setDocumentFinishedLoading):
(WebKit::WebPageLoadTiming::allSubresourcesFinishedLoading const):
(WebKit::WebPageLoadTiming::updateEndOfNetworkRequests):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::WebPageProxy):
(WebKit::WebPageProxy::startNetworkRequestsForPageLoadTiming):
(WebKit::WebPageProxy::endNetworkRequestsForPageLoadTiming):
(WebKit::WebPageProxy::generatePageLoadingTimingSoon):
(WebKit::WebPageProxy::didEndNetworkRequestsForPageLoadTimingTimerFired):
(WebKit::WebPageProxy::didStartProvisionalLoadForFrame):
(WebKit::WebPageProxy::didStartProvisionalLoadForFrameShared):
(WebKit::WebPageProxy::didFinishDocumentLoadForFrame):
(WebKit::WebPageProxy::didLayoutForCustomContentProvider):
(WebKit::WebPageProxy::didReachLayoutMilestone):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::useNetworkLoader):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::dispatchDidStartProvisionalLoad):
(WebKit::WebLocalFrameLoaderClient::dispatchDidFinishDocumentLoad):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):
(WebKit::WebPage::addResourceRequest):
(WebKit::WebPage::removeResourceRequest):
(WebKit::WebPage::dispatchDidReachLayoutMilestone):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/mac/TiledCoreAnimationDrawingArea.mm:
(WebKit::TiledCoreAnimationDrawingArea::sendPendingNewlyReachedPaintingMilestones):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(TEST(WKNavigation, GeneratePageLoadTiming)):
* Tools/TestWebKitAPI/cocoa/TestNavigationDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestNavigationDelegate.mm:
(-[TestNavigationDelegate _webView:didGeneratePageLoadTiming:]):

Canonical link: https://commits.webkit.org/282826@main
